Expressions that modify variables used in bounds declarations.

dtarditi opened the pull request:
We are currently checking that bounds declarations are valid after assignments to variables with bounds declarations.  We also need to check that bounds declarations are valid after assignments to variables _used_ in bounds declarations.   This work is described in issue #502.

This change implements the first part of the work in issue #502.   It computes a mapping from expressions that modify variables to sets of in-scope bounds declarations that use the modified variables.

Determining whether an expression modifies a variable used by an in-scope bounds declaration is a lexically-scoped computation     When checking bounds declarations, we are currently traversing the AST in a lexically-scoped fashion.  We plan to change the traversal of the AST to be based on control flow, so that we can incorporate dataflow facts.   It isn't easy in clang to determine the lexical scope that an expression occurs in.   Clang doesn't have back pointers from expressions to their parent expressions or from statements to enclosing lexical scopes.  That makes it difficult to answer the question of what set of bounds declarations are in scope.  To deal with this, we compute a map that can be consulted at any point in a function.

We compute the information for each function, but do not use it.  For compilation units that don't use Checked C extensions, the maps will be empty.  We want to get coverage on the different kinds of ASTs that we might encounter, so the code is left on during bounds declaration checking.

Testing:
- I wrote a new set of Checked C clang tests that contain expressions that modify variables used in bounds declarations.  I then used this to dump the results of the analysis and manually inspect that the analysis is producing expected results.  The compiler will eventually produce error messages in some cases.   These are in tests/CheckedC/static-checking/bounds-side-effects.c.
- Passes existing Checked C and Checked C clang tests on Windows.
- Passed debug Linux clang tests.
- Other testing in progress.

sxl463 commented on lib/Sema/CheckedCAlias.cpp:
```diff
@@ -333,3 +336,300 @@ void Sema::TrackMemberBoundsDependences(FieldDecl *FD, BoundsExpr *BE) {
+   Sema::BoundsDependencyTracker::DependentMap &Map;
+
+   void AddDependence(VarDecl *D) {
+    if (D == BoundsDecl)   // Do not add self-dependences.
```
Just a personal view, I prefer letting constant variables be left values, like:
BoundsDecl == D;
By doing this we can avoid the equal sing missing bug, buggy code like D = BoundsDecl will never appear again.

sxl463 commented on lib/Sema/SemaBounds.cpp:
```diff
@@ -2759,6 +2759,15 @@ namespace {
+  ModifiedBoundsDependencies Tracker;
+  // Compute a mapping from expressions that modify lvalues to in-scope bounds
+  // declarations that depend upon those expressions.  We plan to change
+  // CheckBoundsDeclaration to traverse function bodies in an order determined
```
Here "traverse function bodies" sounds like an inter-procedural process.

sxl463 commented on lib/Sema/SemaDecl.cpp:
```diff
@@ -12922,13 +12922,19 @@ void Sema::ActOnBoundsDecl(DeclaratorDecl *D, BoundsAnnotations Annots,
   // members to what bounds declarations depend upon them.
   if (FieldDecl *FD = dyn_cast<FieldDecl>(D))
     TrackMemberBoundsDependences(FD, BoundsExpr);
+  else if (VD && !VD->isLocalVarDeclOrParm()) {
```
LLVM code uses "Param" as the abbreviation for parameter. I think we can update "Parm" to "Param". Sounds better.

dtarditi commented:
I have addressed your code review feedback:
- Made a class member const and changed a comparison to have the const value on the left-hand side.
- Changed the wording of a comment for clarity.
